builtInFunctions/esdtModifyRoyalties.go

```diff
@@ -0,0 +1,128 @@
+package builtInFunctions
+
+import (
+	"fmt"
+	"math/big"
+	"sync"
+
+	"github.com/multiversx/mx-chain-core-go/core"
+	"github.com/multiversx/mx-chain-core-go/core/check"
+	vmcommon "github.com/multiversx/mx-chain-vm-common-go"
+)
+
+const (
+	tokenIDIndex      = 0
+	nonceIndex        = 1
+	newRoyaltiesIndex = 2
+)
+
+type esdtModifyRoyalties struct {
+	baseActiveHandler
+	globalSettingsHandler vmcommon.GlobalMetadataHandler
+	storageHandler        vmcommon.ESDTNFTStorageHandler
+	rolesHandler          vmcommon.ESDTRoleHandler
+	accounts              vmcommon.AccountsAdapter
+	funcGasCost           uint64
+	mutExecution          sync.RWMutex
+}
+
+// NewESDTModifyRoyaltiesFunc returns the esdt modify royalties built-in function component
+func NewESDTModifyRoyaltiesFunc(
+	funcGasCost uint64,
+	accounts vmcommon.AccountsAdapter,
+	globalSettingsHandler vmcommon.GlobalMetadataHandler,
+	storageHandler vmcommon.ESDTNFTStorageHandler,
+	rolesHandler vmcommon.ESDTRoleHandler,
+	enableEpochsHandler vmcommon.EnableEpochsHandler,
+) (*esdtModifyRoyalties, error) {
+	if check.IfNil(accounts) {
+		return nil, ErrNilAccountsAdapter
+	}
+	if check.IfNil(globalSettingsHandler) {
+		return nil, ErrNilGlobalSettingsHandler
+	}
+	if check.IfNil(enableEpochsHandler) {
+		return nil, ErrNilEnableEpochsHandler
+	}
+	if check.IfNil(storageHandler) {
+		return nil, ErrNilESDTNFTStorageHandler
+	}
+	if check.IfNil(rolesHandler) {
+		return nil, ErrNilRolesHandler
+	}
+
+	e := &esdtModifyRoyalties{
+		accounts:              accounts,
+		globalSettingsHandler: globalSettingsHandler,
+		storageHandler:        storageHandler,
+		rolesHandler:          rolesHandler,
+		funcGasCost:           funcGasCost,
+		mutExecution:          sync.RWMutex{},
+	}
+
+	e.baseActiveHandler.activeHandler = enableEpochsHandler.IsDynamicESDTEnabled
+
+	return e, nil
+}
+
+// ProcessBuiltinFunction saves the token type in the system account
+func (e *esdtModifyRoyalties) ProcessBuiltinFunction(acntSnd, _ vmcommon.UserAccountHandler, vmInput *vmcommon.ContractCallInput) (*vmcommon.VMOutput, error) {
+	err := checkArguments(vmInput, acntSnd, e.baseActiveHandler)
+	if err != nil {
+		return nil, err
+	}
+	if len(vmInput.Arguments) != 3 {
+		return nil, ErrInvalidNumberOfArguments
+	}
+
+	err = e.rolesHandler.CheckAllowedToExecute(acntSnd, vmInput.Arguments[tokenIDIndex], []byte(core.ESDTRoleModifyRoyalties))
+	if err != nil {
+		return nil, err
+	}
+
+	e.mutExecution.RLock()
+	funcGasCost := e.funcGasCost
+	e.mutExecution.RUnlock()
+	if vmInput.GasProvided < funcGasCost {
+		return nil, ErrNotEnoughGas
+	}
+
+	esdtData, esdtTokenKey, nonce, err := getEsdtDataAndCheckType(vmInput, acntSnd, e.storageHandler)
```

Comment:
this has to work for non dynamic tokens as well, but only if the token is in the callers balance.

Reply:
Done.